homeassistant/components/sonos/switch.py

```diff
@@ -0,0 +1,224 @@
+        """Poll the device for the current state."""
+        await self.async_remove_if_not_available()
+        await self.hass.async_add_executor_job(self.update_alarm)
+
+    def _update_device(self):
+        """Update the device, since this alarm moved to a different player."""
+        device_registry = dr.async_get(self.hass)
+        entity_registry = er.async_get(self.hass)
+        entity = entity_registry.async_get(self.entity_id)
+
+        if entity is None:
+            raise RuntimeError("Alarm has been deleted by accident.")
+
+        entry_id = entity.config_entry_id
+
+        new_device = device_registry.async_get_or_create(
+            config_entry_id=entry_id,
+            identifiers={(SONOS_DOMAIN, self.soco.uid)},
+            connections={(dr.CONNECTION_NETWORK_MAC, self.speaker.mac_address)},
+        )
+        if not entity_registry.async_get(self.entity_id).device_id == new_device.id:
+            _LOGGER.debug("The alarm is switching the sonos player")
+            # pylint: disable=protected-access
+            entity_registry._async_update_entity(
+                self.entity_id, device_id=new_device.id
+            )
+
+    def update_alarm(self):
+        """Update the state of the alarm."""
+        _LOGGER.debug("Updating the state of the alarm")
+        if self.speaker.soco.uid != self.alarm.zone.uid:
+            self.speaker = get_speaker_from_uid(self.alarm.zone.uid, self.data_sonos)
+            self.speaker.available_alarms.add(self.alarm)
+            self._update_device()
+
+        self._is_on = self.alarm.enabled
+        self._attributes[ATTR_ID] = str(self.alarm_id)
+        self._attributes[ATTR_TIME] = str(self.alarm.start_time)
+        self._attributes[ATTR_DURATION] = str(self.alarm.duration)
+        self._attributes[ATTR_RECURRENCE] = str(self.alarm.recurrence)
+        self._attributes[ATTR_VOLUME] = self.alarm.volume / 100
+        self._attributes[ATTR_PLAY_MODE] = str(self.alarm.play_mode)
+        self._attributes[ATTR_SCHEDULED_TODAY] = self._is_today
+        self._attributes[ATTR_INCLUDE_LINKED_ZONES] = self.alarm.include_linked_zones
+
+        self.schedule_update_ha_state()
+
+    @property
+    def _is_today(self):
+        recurrence = self.alarm.recurrence
+        timestr = int(datetime.datetime.today().strftime("%w"))
+        return (
+            bool(recurrence[:2] == "ON" and str(timestr) in recurrence)
+            or bool(recurrence == "DAILY")
+            or bool(recurrence == "WEEKDAYS" and int(timestr) not in [0, 7])
+            or bool(recurrence == "ONCE")
+            or bool(recurrence == "WEEKDAYS" and int(timestr) not in [0, 7])
+            or bool(recurrence == "WEEKENDS" and int(timestr) not in range(1, 7))
+        )
+
+    @property
+    def is_on(self):
+        """Return state of Sonos alarm switch."""
+        return self._is_on
+
+    @property
+    def device_state_attributes(self):
```

Comment:
New standard is `extra_state_attributes`.

Reply:
I'll change it